### PR TITLE
psp-3846 allow adding/editing/deleting file properties to activity.

### PR DESCRIPTION
**Pull request by devinleighsmith**

### CODE REVIEW

File: backend/dal/Helpers/Extensions/EntityExtensions.cs [hunk 1]
```diff
@@ -136,8 +136,11 @@ public static void SetOriginalConcurrencyControlNumber<T>(this T source, DbConte
                 }
                 else
                 {
-                    context.Entry(oldItem).CurrentValues.SetValues(item);
-                    dbItemsMap.Remove(item.Id);
+                    if (updateChildValues)
```
Comment (devinleighsmith) [inline backend/dal/Helpers/Extensions/EntityExtensions.cs:139]: provide more control over the updateChild logic, to only allow updating child associations and not the contents of children.

File: backend/dal/Repositories/ActivityRepository.cs [hunk 2]
```diff
@@ -87,22 +90,26 @@ public PimsActivityInstance Add(PimsActivityInstance instance)
         public PimsActivityInstance Update(PimsActivityInstance instance)
         {
             instance.ThrowIfNull(nameof(instance));
-            GetById(instance.ActivityInstanceId);
-            if (instance.ActivityTemplate != null)
```
Comment (devinleighsmith) [inline backend/dal/Repositories/ActivityRepository.cs:91]: rather then individually blacklisting all of the related properties that shouldn't be updated (which is error prone as it is easy to forget to add a new property here). Decided to change this to a whitelist model instead, where by default only the base entity is updated.

File: frontend/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -204,35 +202,24 @@ const IndeterminateCheckbox = React.forwardRef(
       rest.onChange && !allDataRef && rest.onChange(e);
       const currentSelected = selectedRef?.current ? [...selectedRef?.current] : [];
       if (isHeaderCheck) {
-        setSelected([]);
-      } else {
-        if (isSingleSelect === true) {
-          currentSelected.splice(0, currentSelected.length);
-          currentSelected.push(row.original);
+        if (e.target.checked && !indeterminate) {
```
Comment (devinleighsmith) [inline frontend/src/components/Table/Table.tsx:205]: I was able to greatly simplify this logic. Need to add unit tests here as this logic has been unstable in the past.

File: frontend/src/components/common/GenericModal.tsx [hunk 4]
```diff
@@ -189,6 +190,10 @@ const StyledModal = styled(Modal)`
     .close {
       color: white;
       opacity: 1;
+      font-size: 3rem;
```
Comment (devinleighsmith) [inline frontend/src/components/common/GenericModal.tsx:193]: corrections to the "x" icon on the modal.

File: backend/api/Constants/FileType.cs [hunk 5]
```diff
@@ -10,5 +10,7 @@ public enum FileType
         Acquisition,
         [EnumMember(Value = "research")]
         Research,
+        [EnumMember(Value = "unknown")]
+        Unknown,
```
Comment (FuriousLlama) [inline backend/api/Constants/FileType.cs:14]: I am guessing this is not really an unknown file type but a file type of type "unknown". If that's the case please add some comments so is not as confussing

File: backend/api/Models/Concepts/ActivityInstance/PimsActInstPropFile.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+namespace Pims.Api.Models.Concepts
+{
+    public class PimsActInstPropFile : BaseAppModel
```
Comment (FuriousLlama) [inline backend/api/Models/Concepts/ActivityInstance/PimsActInstPropFile.cs:3]: PimsActInstPropFile reads like an Entity object. Why not the more readable version? Same for the map

File: backend/dal/Repositories/ActivityRepository.cs [hunk 7]
```diff
@@ -87,22 +90,26 @@ public PimsActivityInstance Add(PimsActivityInstance instance)
...
-            {
-                Context.Entry(instance.ActivityTemplate).State = EntityState.Unchanged;
-            }
-            foreach (PimsActivityInstanceDocument activityInstanceDocument in instance.PimsActivityInstanceDocuments)
-            {
-                Context.Entry(activityInstanceDocument).State = EntityState.Unchanged;
-            }
+            var currentActivity = this.Context.PimsActivityInstances
+                .FirstOrDefault(x => x.ActivityInstanceId == instance.Id) ?? throw new KeyNotFoundException();
+            this.Context.Entry(currentActivity).CurrentValues.SetValues(instance);
```
Comment (FuriousLlama) [inline backend/dal/Repositories/ActivityRepository.cs:95]: it is more readable, but I guess it does not update the related properties if we would want that.

File: backend/dal/Repositories/ActivityRepository.cs [hunk 8]
```diff
@@ -135,6 +144,16 @@ public bool Delete(long activityId)
                 this.Context.PimsActivityInstanceNotes.Remove(activityNote);
             }
 
+            foreach (var propAcqFile in instance.PimsActInstPropAcqFiles)
```
Comment (FuriousLlama) [inline backend/dal/Repositories/ActivityRepository.cs:147]: nit, no need for tiny names

File: backend/dal/Repositories/ActivityRepository.cs [hunk 9]
```diff
@@ -135,6 +144,16 @@ public bool Delete(long activityId)
...
+            {
+                this.Context.PimsActInstPropAcqFiles.Remove(propAcqFile);
+            }
+
+            foreach (var propRsrchFile in instance.PimsActInstPropRsrchFiles)
```
Comment (FuriousLlama) [inline backend/dal/Repositories/ActivityRepository.cs:152]: Not necessarily applicable to the links you changed, but would document, leave orphaned documents if that's the only one they are linked too? maybe the same with notes

File: frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx [hunk 10]
```diff
@@ -19,8 +17,11 @@ export interface IDocumentUploadContainerProps {
 
 export const DocumentUploadContainer: React.FunctionComponent<IDocumentUploadContainerProps> = props => {
   const isMounted = useIsMounted();
-  const { getDocumentTypes, getDocumentTypeMetadata } = useApiDocuments();
-  const { retrieveDocumentMetadataLoading } = useDocumentProvider();
+  const {
```
[the remark this answers is not in the document]
Comment (FuriousLlama) [reply frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:20]: that would be great. This changed quite a bit

**Review by FuriousLlama — APPROVED** [review]

File: backend/api/Constants/FileType.cs (hunk 5 above)
Comment (devinleighsmith) [reply backend/api/Constants/FileType.cs:14]: Yes, fair, will do.

File: backend/api/Models/Concepts/ActivityInstance/PimsActInstPropFile.cs (hunk 6 above)
Comment (devinleighsmith) [reply backend/api/Models/Concepts/ActivityInstance/PimsActInstPropFile.cs:3]: Sure, I can remove the Pims bit, add the Model

File: backend/dal/Repositories/ActivityRepository.cs (hunk 7 above)
Comment (devinleighsmith) [reply backend/dal/Repositories/ActivityRepository.cs:95]: yes, that would need to be done explicitly (perhaps more akin to what I'm doing with property updates here).

File: backend/dal/Repositories/ActivityRepository.cs (hunk 8 above)
Comment (devinleighsmith) [reply backend/dal/Repositories/ActivityRepository.cs:147]: Sure, I was following the naming convention of the entity but I don't mind lengthening these.

File: backend/dal/Repositories/ActivityRepository.cs (hunk 9 above)
Comment (devinleighsmith) [reply backend/dal/Repositories/ActivityRepository.cs:152]: Need to think about this one for a bit.

Comment (devinleighsmith) [reply backend/dal/Repositories/ActivityRepository.cs:152]: I think I'm ok with this. the service enforces the business logic to ensure that any related entities (documents, notes) are deleted/disabled when an activity is deleted, whereas this repository is just responsible for making sure that an activity and all of its fk entities are removed properly. The repository shouldn't need to worry about other entities (documents, notes) other then relationships that cause the delete to fail.

File: frontend/src/components/common/GenericModal.tsx (hunk 4 above)
Comment (asanchezr) [reply frontend/src/components/common/GenericModal.tsx:193]: yeah - that icon always seemed a bit small. thanks for fixing

**Pull request merged**